File: python/factors.py
#!/usr/bin/env python3
import math

def factors(n):
    factors = []
    for i in range(1, int(math.sqrt(n)) + 1):
        if(n % i == 0 and i != n):
            factors.append(i)
            converse = int(n / i)
            if(converse != i and converse != n):
                factors.append(converse)
    return sorted(factors)

assert factors(0) == []
assert factors(1) == []
assert factors(2) == [1]
assert factors(3) == [1]
assert factors(4) == [1, 2]
assert factors(5) == [1]
assert factors(6) == [1, 2, 3]
assert factors(7) == [1]
assert factors(8) == [1, 2, 4]
assert factors(9) == [1, 3]
assert factors(10) == [1, 2, 5]
assert factors(11) == [1]
assert factors(12) == [1, 2, 3, 4, 6]
assert factors(13) == [1]
assert factors(28) == [1, 2, 4, 7, 14]

assert sum(factors(6)) == 6
assert sum(factors(28)) == 28

File: scala/factors.scala
import scala.math

def factors(n: Int) = (1 until (math.sqrt(n).toInt + 1)).filter( (i) => n % i == 0).map( i => List(i, n / i)).flatten.toList.sorted
def properDivisors(n: Int) = factors(n).reverse.drop(1).reverse

val factorsTestData = Seq(
    ( 0, Seq())
  , ( 1, Seq(1))
  , ( 2, Seq(1, 2))
  , ( 3, Seq(1, 3))
  , ( 4, Seq(1, 2, 4))
  , ( 5, Seq(1, 5))
  , ( 6, Seq(1, 2, 3, 6))
  , ( 7, Seq(1, 7))
  , ( 8, Seq(1, 2, 4, 8))
  , ( 9, Seq(1, 3, 9))
  , (10, Seq(1, 2, 5, 10))
  , (11, Seq(1, 11))
  , (12, Seq(1, 2, 3, 4, 6, 12))
  , (13, Seq(1, 13))
  , (28, Seq(1, 2, 4, 7, 14, 28))
)

factorsTestData.foreach( n => assert( factors(n._1) == n._2, s"factors(${n._1}) != ${n._2} (actual: ${factors(n._1)})" ))

assert( properDivisors(0) == Seq() )
assert( properDivisors(1) == Seq() )
assert( properDivisors(2) == Seq(1) )
assert( properDivisors(3) == Seq(1) )
assert( properDivisors(4) == Seq(1, 2) )
assert( properDivisors(6) == Seq(1, 2, 3) )

assert( properDivisors(6).foldLeft(0)(_ + _) == 6 )
assert( properDivisors(28).foldLeft(0)(_ + _) == 28 )
